src/Servers.jl

```diff
@@ -130,278 +59,198 @@ function check_rate_limit(tcp;
+        should be allowed per client IP address; requests exceeding the rate limit will be auto-closed
+ - `stream::Bool=false`: whether the handler should operate on an `HTTP.Stream` to read & write directly
+ - `verbose::Bool=false`: whether simple logging should print to stdout for connections handled
 
 e.g.
 ```
-    HTTP.listen() do http::HTTP.Stream
+    HTTP.listen(; stream=true) do http::HTTP.Stream
         @show http.message
         @show HTTP.header(http, "Content-Type")
         while !eof(http)
             println("body data: ", String(readavailable(http)))
         end
-        HTTP.setstatus(http, 404)
-        HTTP.setheader(http, "Foo-Header" => "bar")
+        setstatus(http, 404)
+        setheader(http, "Foo-Header" => "bar")
         startwrite(http)
         write(http, "response body")
         write(http, "more response body")
+        return
     end
 
     HTTP.listen() do request::HTTP.Request
         @show HTTP.header(request, "Content-Type")
-        @show HTTP.payload(request)
+        @show HTTP.payload(request, String)
         return HTTP.Response(404)
     end
 ```
 """
-listen(f, host::IPAddr=Sockets.localhost, port::Integer=8081; kw...) = listen(f, Sockets.InetAddr(host, port); kw...)
-listen(f, host::AbstractString, port::Integer; kw...) = listen(f, parse(IPAddr, host), port; kw...)
-listen(f, host::AbstractString; kw...) = listen(f, string(host); kw...)
+function listen end
 
-function listen(f::Function,
-                host::Union{Sockets.InetAddr, String},
+getinet(host::String, port::Integer) = Sockets.InetAddr(parse(IPAddr, host), port)
+getinet(host::IPAddr, port::Integer) = Sockets.InetAddr(host, port)
+
+function listen(h,
+                host::Union{IPAddr, String}=Sockets.localhost,
+                port::Integer=8081
                 ;
-                ssl::Bool=false,
-                require_ssl_verification::Bool=true,
-                sslconfig::Union{SSLConfig, Nothing}=nothing,
-                pipeline_limit::Int=ConnectionPool.default_pipeline_limit,
-                tcpisvalid::Function=(tcp; kw...)->true,
-                tcpref::Ref=Ref{Base.IOServer}(),
+                stream::Bool=false,
+                sslconfig::Union{MbedTLS.SSLConfig, Nothing}=nothing,
+                tcpisvalid::Union{Function, Nothing}=nothing,
+                tcpref::Union{Ref, Nothing}=nothing,
                 reuseaddr::Bool=false,
-                connectioncounter::Base.RefValue{Int}=Ref(0),
-                kw...)
-
-    if ssl && sslconfig === nothing
-        sslconfig = SSLConfig(require_ssl_verification)
+                connectioncounter::Ref{Int}=Ref(0),
+                ratelimit::Union{Rational{Int}, Nothing}=nothing,
+                reuse_limit::Int=1, readtimeout::Int=0,
+                verbose::Bool=false)
+
+    # If `h` does not accept a stream wrap it with the stream handling function.
+    if !stream
+        h = let f = h; http::Stream -> handle(f, http) end
```

Comment:
I'm not such a huge fan of this change:
1. this code is _really_ confusing
2. it makes it much more difficult to overload with other handlers

Reply:
I agree; I just pushed another idea; please take a look.